harness/determined/exec/tensorboard.py

```diff
@@ -16,50 +17,82 @@ def set_s3_region(bucket: str) -> None:
 
     region = bucketLocation["LocationConstraint"]
-    print(f"export AWS_REGION={region}")
 
+    os.environ["AWS_REGION"] = str(region)
 
-def poll_tensorboard() -> str:
-    task_id = os.environ["DET_TASK_ID"]
-    port = os.environ["TENSORBOARD_PORT"]
-    tensorboard_addr = f"http://localhost:{port}/proxy/{task_id}"
+
+def wait_for_tensorboard(max_seconds: float, url: str, health_check_fn: Callable[[], bool]) -> bool:
+    """Return True if the process successfully comes up before a deadline."""
+
+    deadline = time.time() + max_seconds
 
     while True:
-        time.sleep(5)
+        if time.time() > deadline:
+            print(f"TensorBoard did not find metrics within {max_seconds} seconds", file=sys.stderr)
+            return False
+
+        if health_check_fn():
+            print("TensorBoard process died before reporting metrics", file=sys.stderr)
+            return False
+
+        time.sleep(1)
+
+        try:
+            res = requests.get(url)
+            res.raise_for_status()
+        except (requests.exceptions.ConnectionError, requests.exceptions.HTTPError):
+            continue
 
         try:
-            res = requests.get(f"{tensorboard_addr}/data/plugin/scalars/tags")
             tags = res.json()
-        except Exception:
+        except ValueError:
             continue
 
         # TensorBoard will return { trial/<id> : { tag: value } } when data is present.
         if len(tags) == 0:
+            print("TensorBoard is awaiting metrics...")
             continue
 
         for val in tags.values():
             if len(val):
-                return "TensorBoard contains metrics"
-
+                print("TensorBoard has metrics")
+                return True
 
-def main(args: List[str]) -> None:
-    parser = argparse.ArgumentParser(description="Determined AI Tensorboard Entrypoint")
-    parser.add_argument("command", type=str, choices=["hdfs", "s3", "service_ready"])
-
-    conf = parser.parse_args(args)
 
+def main(args: List[str]) -> int:
     with open("/run/determined/workdir/experiment_config.json") as f:
         exp_conf = json.load(f)
 
-    if conf.command == "s3":
-        if exp_conf["checkpoint_storage"]["type"] == "s3":
-            set_s3_region(exp_conf["checkpoint_storage"]["bucket"])
+    if exp_conf["checkpoint_storage"]["type"] == "s3":
+        set_s3_region(exp_conf["checkpoint_storage"]["bucket"])
+
+    task_id = os.environ["DET_TASK_ID"]
+    port = os.environ["TENSORBOARD_PORT"]
+    tensorboard_addr = f"http://localhost:{port}/proxy/{task_id}"
+    url = f"{tensorboard_addr}/data/plugin/scalars/tags"
+
+    p = subprocess.Popen(["tensorboard", *args])
+
+    killed_subproc = False
+
+    def pass_signal_to_subproc(signum: int, _: Any) -> None:
+        nonlocal killed_subproc
+        if not killed_subproc:
+            killed_subproc = True
+            p.send_signal(signum)
+        else:
+            sys.exit(2)
+
+    signal.signal(signal.SIGTERM, pass_signal_to_subproc)
```

Comment:
question: Are there any other signals we can expect here?

Reply:
I looked into this, and it turns out that docker kill sends SIGKILL by default, which can't be caught.  Frankly, there's not much point in trying to catch anything else either, so I just removed all the signal handling.